pkg/compactor/block_upload.go

```diff
@@ -365,6 +401,75 @@ func (c *MultitenantCompactor) uploadMeta(ctx context.Context, logger log.Logger
 	return nil
 }
 
+func (c *MultitenantCompactor) validateBlock(ctx context.Context, blockID ulid.ULID,
+	userBkt objstore.Bucket, meta metadata.Meta) error {
+	blockDir, err := os.MkdirTemp(filepath.Join(c.compactorCfg.DataDir), "upload")
+	if err != nil {
+		return errors.Wrap(err, "failed to create temp dir")
+	}
+	defer func() {
+		if err := os.RemoveAll(blockDir); err != nil {
+			level.Warn(c.logger).Log("msg", "failed to remove temp dir", "path", blockDir, "err", err)
+		}
+	}()
+
+	var objectCount, objectBytes int64
+	if err := userBkt.Iter(ctx, blockID.String(), func(pth string) error {
+		fname := filepath.Join(blockDir, pth)
+		if pth == uploadingMetaFilename {
+			fname = filepath.Join(blockDir, block.MetaFilename) // rename to final meta.json in temp dir
+		}
+
+		r, err := userBkt.Get(ctx, pth)
+		if err != nil {
+			return errors.Wrap(err, fmt.Sprintf("failed to get object %q from object storage", pth))
+		}
+
+		f, err := os.Create(fname)
```

Comment:
This may fail for chunks file unless we create "chunks" directory first.

Reply:
Yes, thanks for the reminder